0xProject/0x-mesh
# core: implement network ID detection check on startup

_Functional but a work in progress, a few outstanding points (see notes section)._

## Overview

Closes #237 
- Add a `Metadata` collection to MeshDB to store the networkID we are starting up with
- Add a check during startup that checks if a networkID has been set, compares to the one loaded from the environment, and exits if they do not match.
- The detection check (function `initNetworkId`) also stores the network ID during the initial startup (no previous DB)

## Steps for testing:

1. Start mesh on network N (let it run, collect orders, etc)
```
P2P_LISTEN_PORT=<port> ETHEREUM_RPC_URL=<url> ETHEREUM_NETWORK_ID=3 ./mesh
```
2. Kill the process, and restart with a new network ID:
```
P2P_LISTEN_PORT=<port> ETHEREUM_RPC_URL=<url> ETHEREUM_NETWORK_ID=4 ./mesh
```

You should see:
```
{"level":"error","msg":"Mesh previously started on different Ethereum network; switch networks or remove DB","time":"2019-07-23T16:38:09-07:00"}
{"error_string":"expected networkID to be '4', got '3","level":"fatal","msg":"could not initialize app","time":"2019-07-23T16:38:09-07:00"}
```

## Notes
Still a WIP due to me having questions about the following:
- Is there a preferred key for the new collection other than `[]byte("metadata")`? (see [here](https://github.com/hrharder/0x-mesh/blob/development/meshdb/meshdb.go#L55))
- Are there any changes to method names, comments, etc you would like to see?
- Am I handling the mismatched ID case as you would expect? (see [here](https://github.com/hrharder/0x-mesh/blob/development/core/core.go#L278))
